### PR TITLE
Update tests for online expansion

### PR DESCRIPTION
**Pull request by shuo-wu**
longhorn/longhorn#1674

### CODE REVIEW

**Comment by shuo-wu** [comment]
Updated with manual tests

**Review by innobead — COMMENTED** [review]
The manula test is good, and alos we shuould mention this note to the doc of online volume expansion about that it will only take effect after the running volume gets reattached.

File: manager/integration/tests/test_basic.py [hunk 1]
```diff
@@ -2476,52 +2525,83 @@ def test_expansion_canceling(client, core_api, volume_name, pod):  # NOQA
...
+    resp = read_volume_data(core_api, pod_name, "test_file1")
+    assert resp == test_data1
+    resp = read_volume_data(core_api, pod_name, "test_file2")
+    assert resp == test_data2
+
+    # Without a short wait, the volume replicas consider that the directory
+    # created in func fix_replica_expansion_failure still exists then fail
+    # the following expansion request.
+    # Not sure if it's caused by the caching.
+    time.sleep(5)
```
Comment (c3y1huang) [inline manager/integration/tests/test_basic.py:2586]: Looks like an improvement in stability, should we create an issue for this?

**Review by c3y1huang — CHANGES_REQUESTED** [review]

File: docs/content/manual/release-specific/v1.4.0/test-online-expansion.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+**When** Use `dd` to calculate the data checksum for the Longhorn block device..
+
+```
+dd if=/dev/longhorn/vol bs=1M count=5000 status=none | md5sum
+```
+
+*And* Do online expansion for the volume during the calculation.
+
+**Then** The expansion should success.
```
Comment (khushboo-rancher) [inline docs/content/manual/release-specific/v1.4.0/test-online-expansion.md:36]: Can we elaborate this point as something like this `The expansion should succeed and the size of the volume needs to be verified on the attached node.`

File: manager/integration/tests/test_csi.py [hunk 3]
```diff
@@ -376,6 +371,54 @@ def test_csi_offline_expansion(client, core_api, storage_class, pvc, pod_manifes
...
+       and Check the PVC & PV size
+    7. Wait for the copy complete.
+    8. Calculate the checksum for the copied data inside the block volume
+       asynchronously.
+    9. During the calculation, update pvc.spec.resources to expand the volume
+       again.
+    10. Wait for the calculation complete, then compare the checksum.
+    11. Do cleanup: Remove the original `test_data`as well as the pod and PVC.
+    """
+    pass
```
Comment (khushboo-rancher) [inline manager/integration/tests/test_csi.py:395]: Can we skip the test instead of passing it?

**Review by khushboo-rancher — CHANGES_REQUESTED** [review]

File: docs/content/manual/release-specific/v1.4.0/test-online-expansion.md (hunk 2 above)
Comment (shuo-wu) [reply docs/content/manual/release-specific/v1.4.0/test-online-expansion.md:36]: Sure. Let me add one more statement: `The corresponding block device on the attached node is expanded.`

File: manager/integration/tests/test_basic.py (hunk 1 above)
Comment (shuo-wu) [reply manager/integration/tests/test_basic.py:2586]: Not sure. I spent almost one day resolving this but no luck.

**Comment by shuo-wu** [comment]
https://ci.longhorn.io/job/private/job/longhorn-tests-regression/2817/

File: manager/integration/tests/test_basic.py (hunk 1 above)
Comment (shuo-wu) [reply manager/integration/tests/test_basic.py]: The culprit of this wait: https://github.com/longhorn/longhorn/issues/5169#issuecomment-1367202468
Will fix the issue and remove this wait

File: docs/content/manual/release-specific/v1.4.0/test-online-expansion.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+**When** Directly detach and reattach volume 2.
+
+*And* Do online expansion for volume 2.
+
+**Then** The expansion should get stuck. (old engine image + new instance manager)
+
+*And* The engine CR of volume 2 should contain the expansion error.
+
+*And* Then expansion can be cancelled.
```
Comment (c3y1huang) [inline docs/content/manual/release-specific/v1.4.0/test-online-expansion.md:89]:
```suggestion
*And* expansion can be canceled.
```

File: docs/content/manual/release-specific/v1.4.0/test-online-expansion.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+**When** Do live upgrade for volume 1.
+
+*And* Retry online expansion.
+
+**Then** The expansion should get stuck. (new engine image + old instance manager)
+
+*And* The engine CR should contain the expansion error.
+
+*And* Then expansion can be cancelled.
```
Comment (c3y1huang) [inline docs/content/manual/release-specific/v1.4.0/test-online-expansion.md:73]:
```suggestion
*And* expansion can be canceled.
```

**Review by c3y1huang — CHANGES_REQUESTED** [review]

**Review by khushboo-rancher — APPROVED** [review]
LGTM, Thanks @shuo-wu

**Review by c3y1huang — APPROVED** [review]

**Pull request merged**